mesonbuild/wrap/wrap.py

```diff
@@ -129,6 +129,9 @@ def __init__(self, fname: str, subproject: str = ''):
         self.redirected = False
         if self.has_wrap:
             self.parse_wrap()
+            with open(fname) as file:
+                self.hash_of_wrap_file = hashlib.sha256(file.read().encode('utf-8')).hexdigest()
+            self.hash_of_wrap_file_filename = os.path.join(os.path.dirname(self.filename), self.directory, '.meson-subproject-wrap-hash.txt')
         self.directory = self.values.get('directory', self.name)
```

Comment:
Notice how you only store this filename before self.directory is updated to the correct contents.

Reply:
Thanks for pointing that out. This bug caused the hash to fail to be stored anytime the `.wrap` set `directory` to something other than what the default would be. It was also failing for promoted subprojects. To fix these issues, I've refactored `update_hash_cache()` to accept a string argument for the path to the subproject's configured directory, since that's not available on the `PackageDefinition` class but it is for the `Resolver` class.